### PR TITLE
V10: Migrate member properties to columns on the member table

### LINKED ISSUE #11931: Member last login date not being set
### Which *exact* Umbraco version are you using? For example: 9.0.1 - don't just write v9

8.15.3/8.17.1/9.2.0

### Bug summary

In Umbraco 8.1.5, our member login functionality used to set the last login property of the member. Since upgrading to 8.15.3, this no longer happens (unless the user has first had a failed login attempt).

I have been able to reproduce this behaviour on 8.17.1 and 9.2.0, though on 9.2.0 I can't get the last login date to be set even following a failed login attempt.

### Specifics

We have a login method for members that calls `Membership.ValidateUser()`. This results in a call to `UmbracoMembershipProvider.PerformValidateUser()`.

In [8.1.5](https://github.com/umbraco/Umbraco-CMS/blob/release-8.1.5/src/Umbraco.Web/Security/Providers/UmbracoMembershipProvider.cs#L599), this method would always result in a call to `MemberService.Save()` (where the version is >= 7.3.0.0).

In [8.15.3](https://github.com/umbraco/Umbraco-CMS/blob/release-8.15.3/src/Umbraco.Web/Security/Providers/UmbracoMembershipProvider.cs#L609), this only happens if a 'full save' is required. Otherwise, `MemberService.SetLastLogin()` is called instead.

It looks like this has been the case since v8.6.5, when the following PR was merged to resolve some perf issues in this area: https://github.com/umbraco/Umbraco-CMS/pull/8525

`MemberService.SetLastLogin()` calls `MemberRepository.SetLastLogin()`, in which a SQL query is run to retrieve the ID of the umbracoMemberLastLogin property for the user, which is then used to update its value. For members that don't have a value set for this property, no date is set.

If a 'full save' is forced (i.e. by providing an incorrect password during login) then the next login will cause the last login date to be set (in v8).

In v9, I see the same behaviour re: the last login date not being set on login, but can't set it via an unsuccessful login first either.

I've been able to work around the issue by explicitly setting a last login date on members when they're registered (via `MemberService.CreateMember()`), but wonder if it's something that should be fixed, particularly as it did work prior to 8.6.5.

I'm wondering if we can check whether the select query returns an ID before doing the update and, if not, inserting a row into cmsPropertyData. Though I'm not sure if that's a) possible or b) going to cause a big perf issue in an area that perf was originally an issue hence the change that was made!

### Steps to reproduce

- Install Umbraco 8.1.5
- Run installer and use starter site with SQL Server Compact db
- Create register/login forms. Register uses IMemberService.CreateMember(), login uses Membership.ValidateUser()
- Register as user
- Login as user
- Observe that last login date is set

- Run same steps as above for Umbraco 8.15.3/8.17.1
- After logging in as user, last login date is not set
- Last login date can be set by first providing an incorrect password, before then providing the correct password

- Install Umbraco 9.2.0 (with --SqlCE flag)
- Run installer and use SQL Server Compact db
- Install 'The Starter Kit' package
- Create register/login forms. Register uses IMemberService.CreateMember(), login uses IMemberSignInManager.PasswordSignInAsync()
- Register as user
- Login as user
- Observe that last login date is not set
- Unsuccessfully login as user
- Login as user
- Observe that last login date is still not set

### Expected result / actual result

**Expected**: Last login date is set when a member logs in, irrespective of whether they've already got a value set or not

**Actual**: Last login date is only set in v8 where a member first has a failed login attempt. Last login date is seemingly not set at all in v9.

---
_This item has been added to our backlog [AB#17125](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/17125)_

### PR DESCRIPTION
**Pull request by nikolajlauridsen**
This migrates most of the default Umbraco properties on the member types to be saved directly on the member row in the `cmsMember` table, these properties are:

* umbracoMemberFailedPasswordAttempts
* umbracoMemberApproved
* umbracoMemberLockedOut
* umbracoMemberLastLockoutDate
* umbracoMemberLastLogin
* umbracoMemberLastPasswordChangeDate

These properties have also been moved in the backoffice so they're now in the Member content app instead of content. 

A thing worth mentioning here is that by default all these properties were marked as sensitive, so, for now, these are hardcoded as sensitive in the member content app, we may want to consider a way to configure this at some point. 

Additionally, the date properties are now nullable since they're initialized to null (for instance when you've never logged in), instead of them being `DateTime.MinValue`, these properties are:
* LastLoginDate
* LastPasswordChangeDate
* LastLockoutDate

Lastly, I've re-used the user translations for the properties instead of hardcoded labels, and I've added the `stateApproved` translation because that was the only missing one. 

This also fixes #11931 for V10

## Breaking changes

This PR is breaking because you can no longer access the properties listed above through the `IMember.Properties`, collection, you must now access them through their specific properties I.E.  `IMember.IsLockedOut`. 

Additionally, when previously you resolved a member as published content, all the default properties would be there twice, for instance, `IsLockedOut` would be there both as a property with the alias `umbracoMemberLockedOut` and with the alias `IsLockedOut`, now it'll only be there once, with the alias being the name of the property, so `IsLockedOut` in this instance.

Lastly the nullable dates on a user, I.E. `LastLoginLate` will now be null instead of `DateTime.MinValue` when getting a user with the user service.

## Testing

⚠️ Front end files have updated, remember to run `npm run build` before testing ⚠️

In general members and users should work as before, some things that might be a good idea to test:

* Test the migration, with custom member types and member groups
* Ensure you can CRUD members both in the frontend and backoffice
* Ensure that properties like `LastLoginDate` updates correctly

### CODE REVIEW

**Review by elit0451 — COMMENTED** [review]
The bug we discussed happens when I edit the `Member` member type to include a new group with several custom properties that have different options enabled (_Show on member profile, Member can edit, Is sensitive data_). _An example can be seen in the image below_. You fill in the fields on the members you have in the Members section. Then after the migration, all the custom props on those members will be set as sensitive. You can verify that either on the member type or try to login with a user that has access to the Members section but is not part of the Sensitive data group.

image


Other than that, I didn't find anything peculiar. Also, left a few minor suggestions 😊

File: src/Umbraco.Infrastructure/Persistence/NPocoSqlExtensions.cs [hunk 1]
```diff
@@ -437,6 +455,24 @@ public static Sql<ISqlContext>.SqlJoinClause<ISqlContext> LeftJoin<TDto>(this Sq
             string alias = null) =>
             sql.SqlContext.SqlSyntax.LeftJoinWithNestedJoin<TDto>(sql, nestedJoin, alias);
 
+        /// <summary>
+        /// Appends an LEFT JOIN clause using a nested query.
+        /// </summary>
+        /// <param name="sql">The SQL statement</param>
+        /// <param name="nestedSelect">The nested sql query</param>
+        /// <param name="alias">An optional alias for the joined table.</param>
+        /// <returns>A SqlJoin statement</returns>
```
Comment (elit0451) [inline src/Umbraco.Infrastructure/Persistence/NPocoSqlExtensions.cs:464]:
```suggestion
        /// <param name="sql">The SQL statement.</param>
        /// <param name="nestedSelect">The nested sql query.</param>
        /// <param name="alias">An optional alias for the joined table.</param>
        /// <returns>A SqlJoin statement.</returns>
```

File: src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs [hunk 2]
```diff
@@ -218,9 +211,80 @@ public IEnumerable<ContentPropertyDisplay> MapMembershipProperties(IMember membe
                     Config = new Dictionary<string, object>
                     {
                         { "IsRequired", true }
+                    },
+                },
+
+                // These properties used to live on the member as property data, defaulting to is sensitive, to we set them to sensitive here too
```
[the remark this answers is not in the document]
Comment (nikolajlauridsen) [reply src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs:217]: Yeah that was quite the word salad, much clearer now, thanks 😄

**Comment by nikolajlauridsen** [comment]
Great catch on the bug 💪 

Turns out it was a bug that goes back to V9, but it's fixed now in this branch 😄

I've also made a PR for V9 with the same fix #12282

**Comment by elit0451** [comment]
Looks good, gave it a final test and didn't find any other issues 💪

**Pull request merged**